aiperf/workers/worker.py

```diff
@@ -147,6 +163,246 @@ async def _handle_profile_cancel_command(
+                phase=message.phase,
+                delayed_ns=None,  # TODO: set this properly (from record if available?)
+            )
+            if self.is_trace_enabled:
+                self.trace(f"Returning credit {return_message}")
+            # NOTE: Do not do this execute_async, as we want to give the credit back as soon as possible.
+            await self.credit_return_push_client.push(return_message)
+
+    async def _execute_single_credit_internal(self, message: CreditDropMessage) -> None:
+        """Run a credit task for a single credit."""
+        drop_perf_ns = time.perf_counter_ns()  # The time the credit was received
+
+        if not self.inference_client:
+            raise NotInitializedError("Inference server client not initialized.")
+
+        # retrieve the prompt from the dataset
+        conversation_response: ConversationResponseMessage = (
+            await self.conversation_request_client.request(
+                ConversationRequestMessage(
+                    service_id=self.service_id,
+                    conversation_id=message.conversation_id,
+                    credit_phase=message.phase,
+                )
+            )
+        )
+        if self.is_trace_enabled:
+            self.trace(f"Received response message: {conversation_response}")
+
+        # Check for error in conversation response
+        if isinstance(conversation_response, ErrorMessage):
+            await self._send_inference_result_message(
+                RequestRecord(
+                    model_name=self.model_endpoint.primary_model_name,
+                    conversation_id=message.conversation_id,
+                    turn_index=0,
+                    turn=None,
+                    timestamp_ns=time.time_ns(),
+                    start_perf_ns=time.perf_counter_ns(),
+                    end_perf_ns=time.perf_counter_ns(),
+                    error=conversation_response.error,
+                )
+            )
+            return
+        turn_list = []
+        for turn_index in range(len(conversation_response.conversation.turns)):
+            turn = conversation_response.conversation.turns[turn_index]
+            turn_list.append(turn)
+            # TODO: how do we handle errors in the middle of a conversation?
+            record = await self._build_response_record(
+                conversation_response, message, turn, turn_index, drop_perf_ns
+            )
+            await self._send_inference_result_message(record)
+            resp_turn = await self._process_response(record)
+            if resp_turn:
+                turn_list.append(resp_turn)
+
+    async def _build_response_record(
+        self,
+        conversation_response: ConversationResponseMessage,
+        message: CreditDropMessage,
+        turn: Turn,
+        turn_index: int,
+        drop_perf_ns: int,
+    ) -> RequestRecord:
+        record = await self._call_inference_api_internal(message, turn)
```

Comment:
i think ideally this one line can be extracted out of here and moved as the first line in `_send_api_request_and_process_response` (assuming this doesn't complicate things a ton with regards to parameters.

Reply:
Delaying as well.